trieste/models/optimizer.py

```diff
@@ -0,0 +1,133 @@
+"""
+
+
+@dataclass
+class Optimizer:
+    optimizer: Union[gpflow.optimizers.Scipy, tf.optimizers.Optimizer]
+    minimize_args: Dict[str, Any] = field(default_factory=lambda: {})
+    compile: bool = False
+
+    def create_loss(self, model: tf.Module, dataset: Dataset) -> LossClosure:
+        x = tf.convert_to_tensor(dataset.query_points)
+        y = tf.convert_to_tensor(dataset.observations)
+        data = (x, y)
+        return create_loss_function(model, data, self.compile)
+
+    def optimize(self, model: tf.Module, dataset: Dataset) -> OptimizeResult:
+        loss_fn = self.create_loss(model, dataset)
+        variables = model.trainable_variables
+        return self.optimizer.minimize(loss_fn, variables, **self.minimize_args)
+
+
+@dataclass
+class TFOptimizer(Optimizer):
+    max_iter: int = 100
+    batch_size: Optional[int] = None
+    dataset_builder: Optional[DatasetTransformer] = None
+
+    def create_loss(self, model: tf.Module, dataset: Dataset) -> LossClosure:
+        def creator_fn(data: Union[Tuple, Iterable]):
+            return create_loss_function(model, data, self.compile)
+
+        if self.dataset_builder is None and self.batch_size is None:
+            x = tf.convert_to_tensor(dataset.query_points)
+            y = tf.convert_to_tensor(dataset.observations)
+            return creator_fn((x, y))
+        elif self.dataset_builder is None:
+            d = tf.data.Dataset.from_tensor_slices((dataset.query_points, dataset.observations))
+            size = len(dataset)
+            data = iter(
+                d.shuffle(size)
+                .batch(self.batch_size)
+                .prefetch(tf.data.experimental.AUTOTUNE)
+                .repeat()
+            )
+            return creator_fn(data)
+
+        data = self.dataset_builder(dataset, self.batch_size)
+        return creator_fn(data)
+
+    def optimize(self, model: tf.Module, dataset: Dataset) -> None:
+        loss_fn = self.create_loss(model, dataset)
+        variables = model.trainable_variables
+
+        @jit(apply=self.compile)
+        def train_fn():
+            self.optimizer.minimize(loss_fn, variables, **self.minimize_args)
+
+        for _ in range(self.max_iter):
+            train_fn()
+
+
+@singledispatch
+def create_optimizer(optimizer, optimizer_args: Dict[str, Any]) -> Optimizer:
+    pass
+
+
+@create_optimizer.register
+def _create_tf_optimizer(
+    optimizer: tf.optimizers.Optimizer,
+    optimizer_args: Dict[str, Any],
+) -> Optimizer:
+    return TFOptimizer(optimizer, **optimizer_args)
+
+
+@create_optimizer.register
+def _create_scipy_optimizer(
+    optimizer: gpflow.optimizers.Scipy,
+    optimizer_args: Dict[str, Any],
+) -> Optimizer:
+    return Optimizer(optimizer, **optimizer_args)
+
+
+Data = Union[Tuple[tf.Tensor, tf.Tensor], Iterable[Tuple[tf.Tensor, tf.Tensor]]]
+
+
+@singledispatch
+def create_loss_function(model, dataset: Data, compile: Optional[bool] = False) -> LossClosure:
+    pass
```

Comment:
would it be helpful to explicitly raise a NotImplementedError or TypeError or so in case someone passes a model that happens to not inherit from either Internal- or ExternalData...Mixin?